### ISSUE TITLE
Feature request: extending external configs

### ISSUE THREAD

**Issue by raveclassic** [editorconfig/editorconfig#236]
I suggest adding some kind of rule to allow extending external configurations like [it's done in eslint](http://eslint.org/docs/user-guide/configuring#extending-configuration-files).

For example, I have different configurations in external npm package and as editorconfig is related to codestyle like eslint it would be great to reuse the configs in every project.

**Comment by DullReferenceException** [comment]
+1 for this. It'd be very helpful for enforcing organizational coding styles.

**Comment by thomasklein** [comment]
+1

**Comment by gilmrjc** [comment]
Quoting the specification:

> When opening a file, EditorConfig plugins look for a file named .editorconfig in the directory of the opened file and in every parent directory. A search for .editorconfig files will stop if the root filepath is reached or an EditorConfig file with root=true is found.

You can have a global .editorconfig in you home directory and extend it in your project dir, and also have a editorconfig file in every subdir of your project. I guess this is the closes thing to the feature you request.

**Comment by raveclassic** [comment]
@gilmrjc This does not solve the problem. I need to extend `.editorconfig` from an npm-package which is bundled to the project. For now the only way is to copy-paste it from project to project.

**Comment by rasdaniil** [comment]
Can't bundle it together with tslint settings because of that :(

**Comment by kirillgroshkov** [comment]
Still needed!

**Comment by hax** [comment]
I assume editorconfig settings are much simpler than eslint/tslint, and `extends` make it a little bit hard to see all settings value, so maybe we should always copy the file?

**Comment by raveclassic** [comment]
@hax Exstensibility is _always_ better than copy-pasting

**Comment by xuhdev** [comment]
@editorconfig/board-member Surprisingly we didn't have much discussion on this... I personally think this can be a good extension allowing large projects to modulize their `.editorconfig` configurations if `extend` is replaced by `include`. For example, we/user can define some standard coding styles and they can be included directly. Let's say we define two coding styles: `./editorconfigs/c-gnu.editorconfig` and `./editorconfigs/c-kr.editorconfig`. Then, in different sub source tree, we can use

```
[*.c]
include = ../.editorconfig/c-gnu.editorconfig
[*.cc]
include = ../.editorconfig/c-kr.editorconfig
```

to switch between different coding styles. However, I think this should be motivated for languages other than C/C++ as well, since I haven't seen another language using significantly standard different coding styles. Thoughts?

**Comment by ffes** [comment]
In general I like the idea.

But what about security? Are absolute paths allowed, or required? Is `..` allowed? If so, how many levels deep? Should be able to get out of the root of the project? Is it even possible to really detect the root of a project?

One might say that we only have some basic settings that go into a `.editorconfig` file why worry about security. Think again, see https://github.com/sgur/vim-editorconfig/issues/31 for instance.

And since it is possible to add there own tags, like for instance [the Roslyn compiler for .Net](https://docs.microsoft.com/en-us/visualstudio/ide/editorconfig-code-style-settings-reference), basically the sky is the limit. So security should always something to think about. Although you can say that when somebody adds his own tags it is their responsibility to think about security.

Anyway, we should be very specific about what is allowed and what not. Remember #360 / #361

**Comment by xuhdev** [comment]
@ffes That's a true concern: We will get the system overly complicated. If we are not motivated enough, I would prefer not supporting `include`. Are you aware of some template language can generate files following these kinds of rules? Perhaps OP can use it to regenerate `.editorconfig` files: effectively `include` and `extend` are supported this way.

**Comment by mrmckeb** [comment]
Hi all, it looks like this hasn't progressed. 

I'd be happy to help out. It looks like `extends` will work for most people.

**Comment by sorenhoyer** [comment]
@mrmckeb are you going to look into this?

**Comment by mrmckeb** [comment]
Hi @sorenhoyer, I was waiting for a response from the team - I can still see this being useful for many people.

**Comment by muehre** [comment]
I do agree. Pretty useful especially for sharing editor configs between multiple projects.
Would be awesome to extend a company preset of editorconfig

**Comment by florianb** [comment]
@xuhdev, @ffes  - Since interest in this problem is continuing, I would like to propose that we may open a formal discussion about a include-functionality.

I am very curious about the intended use-cases since i think it will be a bit of work to create a solution which is performant, safe and meeting a users expectation.

In my current opinion there is (at least) one EC-file in the repository which should be configured the way the repo-owners want their code being formatted. Having external dependencies which might change eventually may disrupt the repo's intentional formatting.

**Comment by xuhdev** [comment]
I agree -- do you think it would solve the concern by limiting the included file to be no higher than the directory level of the including `.editorconfig` file?

**Comment by florianb** [comment]
I don't know (yet) - as @ffes mentioned i'd fear a directory-traversal attack.

Currently my main pain-point is that i don't see a case where a repo-owner would prefer implicit over explicit EC-rules. Because that's probably the reason why most of the repos (i worked with) have a `root = true`-statement.

---
This could be a starting-point for further research:

- https://en.wikipedia.org/wiki/Directory_traversal_attack
- https://www.owasp.org/index.php/Path_Traversal

**Comment by stevenvachon** [comment]
...

**Comment by mrmckeb** [comment]
@florianb I think there are some sensible defaults that I'd like to see enforced. This is particularly valuable if you work in an organisation with many projects.

I don't think this is hard to set up per project, but it would be great to inherit company defaults - for both new projects and updates to existing projects. It also ensures consistency.

There are approaches we could take as an organisation of course, like populating the file with defaults via a custom package, and updating files (trying to take into account user changes).

Does anyone have any other ideas on how this could be solved without having config extension?

**Comment by cxw42** [comment]
I can think of two options for various use cases.  In the following, `/` is the repo route.

- For the corporate-standards use case, this could be a plug-in option rather than something in a .editorconfig file.  E.g., after the normal directory search finishes, the plug-in can also check one or more files the user (or IT Dept.) has configured.  I would suggest pulling the list from an environment variable or a dotfile in a known location (`%appdata%` or `~/.config`) rather than a project for, for ease of corporate administration.

- For the npm-packaged rules use case, symlink `/.editorconfig -> /node_modules/foo/.editorconfig`.  Put project-specific EC files, and source files, in `/src`.

Requesters of this feature, what say you?  Neither of these ideas is a panacea, but both avoid requiring plugins to implement a system as complex as eslint's.

**Comment by mrmckeb** [comment]
I think those are great suggestions, and we'll likely go with the second as we don't configure developer environments to that level.

**Comment by vith** [comment]
> * For the npm-packaged rules use case, symlink `/.editorconfig -> /node_modules/foo/.editorconfig`. Put project-specific EC files, and source files, in `/src`.

Do symlinks committed to git from a unix system work properly when checked out on a windows system and vice versa?

Trying to answer my own question, it seems to depend on configuring windows and git to make it work: https://stackoverflow.com/a/59761201

I wonder what the situation is with version control systems other than git as well.

**Comment by cxw42** [comment]
@vith Those are good questions --- I unfortunately do not know for sure.  I do know that symlinks work fine for me on Cygwin running in Windows, which is my primary npm environment.

**Comment by mrmckeb** [comment]
I'd argue that any developer on Windows should try to use WSL if they can, but I don't think we can say that WSL/Cigwin/etc is a solution to that problem sadly.

So perhaps that brings us back to the config question. Could we overcome concerns of a directory-traversal attack? It seems to me that the implementations (editor plugins, etc.) themselves would need to handle this, rather than the config.

**Comment by florianb** [comment]
@mrmckeb thank you very much - i see your point. And absolutely valid, the implementations should disallow misuse as a last resort.

But that's a new issue - i think EditorConfig should be way more strict and explicit about it's implementation. And thinking about past discussions this is very hard to achieve since we have a multitude of different platforms with different abilities. Currently it seems we found some kind of sweet spot where most of the tools using EditorConfig are actually able to support its feature set.

And i am sorry i have to say it like this for my perspective: we (or at least some of us) are thinking hard about ways to improve the standard to better support and represent gaining requirements. But it takes its time to do that right.

---

To come back to the issue - what is actually the issue? I became mislead by "request[ing a feature for] extending external configs" which is already an solution. For a lot of reasons that's not easy to do.

But the further discussion showed the real pain point is, how provide a standardized editorconfig to a group of users? And this should really not be so much of a problem.

To answer @raveclassic first intent: Why not adding [`postinstall` & `preuninstall` hooks](https://docs.npmjs.com/misc/scripts) to your script which ensures the symlink is set to your module's folder?

These hooks are available on many platforms and package managers.

If you drive an organization you already have to provide a dev-environment, don't you (i do)? How do you provide settings and tooling at all?

**Comment by vith** [comment]
I don't understand the concern around being able to extend a config outside the current directory. It seems like a desirable capability, i.e. in a monorepos where the .editorconfig might live in a sibling directory next to the code packages.

What could make directory traversal an "attack" in our context?

**Comment by mrmckeb** [comment]
No problem @florianb, using a `postinstall` hook was what I planned to do for our situation - but again, here we also need to cater for how different operating systems might handle that... which is tricky.

We know that there are users that have directories in Windows, but use WSL or Docker to run projects. In this example, only one of those environments would get the symlink.

Further, the root config would not be extendable (at root). Only on a per-folder basis. This may cause issues for some people.

Other configs, such as ESLint, Prettier, Stylelint, TypeScript, etc. allow extending and I feel this is the path forward for EditorConfig too.

**Comment by florianb** [comment]
@mrmckeb for the example of NPM i'd stick with a node-script creating the symlinks. We use root-configs in a per-org/provider-manner. For example:

```
~/
  .editorconfig (my personal preferences)
  github/
    project1/
      .editorconfig (project-preferences)
    project2/
  myorg_1/
    .editorconfig (org-preferences)
    project1/
    project2/
```

EditorConfig is not really comparable to ESLint (, Prettier, Stylelint, TypeScript, etc.) - they live in a strict environment with more or less predictable foundations. EditorConfigs range is far greater than that.

The implications for changing EditorConfigs core behavior are very hard to predict. Handling paths in a safe manner for all environments is nearly impossible. And since EditorConfig rules are like being used by file-manipulating tools i expect it being a possible attack surface for [directory traversal attacks](https://en.wikipedia.org/wiki/Directory_traversal_attack).

We might not imagine that now, but after implementing it, after some weird tool-chain at the other end of the EditorConfig universe downstreams that change it might lead to an attack.

And i think it's part of my obligations to consider that and compare the estimated efforts of maintaining such a solution with the efforts to solve your problem in another way.

I understand this is completely frustrating because it seems like i'm against adding an option to extend the EditorConfig at all. I am not. But currently i am not convinced it's worth taking the amount of work and risk against adding an additional EditorConfig file or symlink on your platforms.

**Comment by cxw42** [comment]
I agree that eslint is not directly comparable.  eslint is a single program, and runs in a relatively controlled environment (an npm project tree).  EditorConfig is a standard, implemented by dozens of programs running in arbitrary environments (e.g., `/etc`).

I think this relates to the general question in #415.  Currently, the most basic benefit of `.editorconfig` files is that they reduce the chance of whitespace deltas in diffs.  Any configuration from outside a `root=true` repo risks negating that benefit.  If `.editorconfig` files are going to broaden their scope, then an include/extends mechanism might be worth the tradeoff.

**Comment by florianb** [comment]
>  Any configuration from outside a `root=true` repo risks negating that benefit

I don't think so, the most recent setting wins and the EditorConfig files are read from top to bottom. So any whitespace-setting should remain if set.

I absolutely agree - EditorConfig must evolve and we must take new toolchains into considersation.

Hopefully everybody using this project is adding some thoughts or feedback on #415!

**Comment by snebjorn** [comment]
@florianb 
> Currently my main pain-point is that i don't see a case where a repo-owner would prefer implicit over explicit EC-rules. Because that's probably the reason why most of the repos (i worked with) have a `root = true`-statement.

Not sure if this is still a concern of yours. But I would argue that the success of https://github.com/prettier/prettier strongly suggests that people prefer implicit (zero config) formatting.

If you're unfamiliar with Prettier then it's an highly opinionated code formatter that basically have become the standard in the javascript/typescript world. And it's awesome :) No more format discussions, no more inconsistent formatting ❤️

Another thing. I've seen NPM packages mentioned a few times. But don't forget about the other package managers like, pip, nuget, maven, etc.. 
My point is that the implementation shouldn't be biased towards NPM

**Comment by florianb** [comment]
Thanks @snebjorn - that's a valid point. And i just want to make one thing clear, i am not against adding extend/include - i just point out how this behavior accomplished with already existing functionality. And it is of course questionable if my opinion applies to everybody's workflow.  

I'm familiar with prettier and i'm embracing JustWorks:tm: software by myself. Also i picked NPM as an example due to the previous discussion - i know a lot of package-managers and most of them would allow to define post-installation actions.

I am pointing out that it **might** not be worth to take the effort implementing this functionality taking the risk of a traversal vulnerability and creating probably overlapping functionality.

Since this issue gains constant traction it should probably just go into voting?

I have no clue when we consider an issue having a significant amount of demand...

**Comment by cxw42** [comment]
Based on the above, my understanding is that we are looking for a way of reading .editorconfig files outside a repo's tree without using symlinks (which may be dicey on vanilla Windows).  I note that the discussion above has been about specific files.  However, .editorconfig files are currently identified by the _directory that contains them_.  What if we built on that foundation?

I propose a top-level property called `search` that would add a directory to the plugin's search path. 
 Then a malicious .editorconfig file would not be able to expressly ask for `/etc/passwd`.  Using `search` would leave the `include` and `extend` keywords available for possible future use.  `search` would be orthogonal to `root`.

**First-draft specification text**

```diff
 ## File Processing
 ...
 The search shall stop if an EditorConfig file is found 
 with the root key set to true in the preamble or when 
 reaching the root filesystem directory.
+
+If the preamble includes a 'search' key specifying a path, 
+that path will be resolved to an absolute path with no
+symlinks.  Any path including a loop will be discarded.
+The resolved path will be added at the end 
+of the list of directories searched for EditorConfig files
+if that path exists and is readable by the current user.
```
(the basics.  You can add to the plugin's search path.  Resolving symlinks right up front makes the rest of the processing simpler.  Detecting loops at this stage reduces DoS risk.)
```diff
+Multiple 'search' keys may be given.  The paths are 
+searched in order the 'search' keys occur in the file.
+However, if a particular resolved path has already been
+added to the list of paths, it will not be added again.
```
(a clear rule in case of unintentional duplicates.)
```diff
+While processing EditorConfig files in 'search' directories, 
+ all 'search' keys in those files shall be ignored.
```
(prevent a DoS attack due to combinatorial explosion in malicious `.editorconfig` files.)

What say you all?

**Comment by florianb** [comment]
This sounds good!

I'd probably not use the term "resolved" as in many frameworks this is used to describe a functionality to literally resolve an absolute path, containing dots and stuff, to an absolute path.

I'd like to propose making this more strict like any path must be relative and must consist of real directories (preventing links and linking symbols like single/double-dot).

In order to implement that safely i guess it will be necessary to create a safe resolve function and not to pass the string to any os-method directly.

**Comment by cxw42** [comment]
@florianb sure - let's use "canonicalize" instead, since that is what the OS/library facilities refer to. 
 E.g., <http://man7.org/linux/man-pages/man3/realpath.3.html>, <https://docs.microsoft.com/en-us/windows/win32/api/pathcch/nf-pathcch-pathcchcanonicalize>.  Canonicalizing is a well-studied problem, and existence/readability checks will add additional safety.  Nothing is 100% secure, but OS/library functions have much more testing behind them than any individual EC plugin does :) .

**Comment by Kasp42** [comment]
+1

**Comment by snowinmars** [comment]
> I don't understand the concern around being able to extend a config outside the current directory. It seems like a desirable capability, i.e. in a monorepos where the .editorconfig might live in a sibling directory next to the code packages.
> 
> What could make directory traversal an "attack" in our context?

Agreed: there are a lot of valid cases when a user what to extend a config outside of the current directory.
I came here due to I have a monorepo, microservice per folder. I definitely want to apply .editorconfig from the .git root to a project, and it's possible only if a project folder contains .editorconfig that extends root config:

- .git
- .editorconfig # root
- svc1
- - .editorconfig # extend root
- - ... files
- svc2 # opening this folder in, f.e., Rider should apply root .editorconfig
- - .editorconfig # extend root
- - ... files

So, yea.

Otherwise, directory traversal risk is a stuff that no one wants to apply. I do expect some startup IDE not knowing how to deal with it.

**Comment by itsezc** [comment]
Has there been any movement on this issue? I think shareable configs are widespread, think IDE or lint configurations. I mainly use TypeScript & Rust, and hate creating an editorconfig for every project and maintaining it, and having a shareable config makes life so much easier, especially when working across multiple projects with the same tooling / stack.

**Comment by TheXenocide** [comment]
Just wanted to add that we have a desire to include a base .editorconfig in a NuGet package or custom project SDK (`<Project Sdk="Package.Id">` in a .csproj) which gets restored to a non-relative path controlled by tooling, so it might be nice for a formal explanation for how to use tooling-based paths (e.g. `extends: nuget:Some.Package.Id` or some such), at least enough to provide some degree of consistency for downstream implementations to extend.

More discussion here: https://github.com/dotnet/roslyn/issues/19028

**Comment by rendevor** [comment]
Is there any progress on this? 
Is it possible to include HTTP as a source and ping a particular version during implementation? 
Should be helpful for huge companies across organizations and repositories.

**Comment by TheXenocide** [comment]
We are also still looking for better solutions for this problem. 

We have many separate repositories in multiple separate VCS solutions and really don't want to have to copy/paste the same file to some 50+ repositories every time we make a change to it. Automating this is a non-trivial task, provides a sub-par experience (and limits local override flexibility, to a lesser degree). 

It seems like Microsoft is reluctant to add support for this functionality to their editors (VS, VSCode) without the underlying spec having some sort of formal agreement on it, so we're dependent on this in order to continue lobbying for the downstream support we need.

**Comment by Piedone** [comment]
If you're working with .NET, then you can do quasi-inheritance/extension with .globalconfig (which is a superset of .editorconfig) files and their `global_level` property. This is what we also do in our [Lombiq .NET Analyzers project](https://github.com/Lombiq/.NET-Analyzers) to have cascading configuration. E.g., [this file with `global_level = 0`](https://github.com/Lombiq/.NET-Analyzers/blob/dev/Lombiq.Analyzers/Lombiq.Analyzers.globalconfig) has the base configuration, and [this other one with `global_level = 10`](https://github.com/Lombiq/.NET-Analyzers/blob/dev/Lombiq.Analyzers.OrchardCore/Lombiq.Analyzers.OrchardCore.globalconfig) overrides some of it.

**Comment by ffes** [comment]
> It seems like Microsoft is reluctant to add support for this functionality to their editors (VS, VSCode) without the underlying spec having some sort of formal agreement on it, so we're dependent on this in order to continue lobbying for the downstream support we need.

Microsoft don't even follow the official specs in VS, especially `trim_trailing_spaces`. And the VSCode EditorConfig plugin is not maintained by Microsoft, but by [volunteers in this organisation](https://github.com/editorconfig/editorconfig-vscode).

**Comment by TheXenocide** [comment]
> If you're working with .NET, then you can do quasi-inheritance/extension with .globalconfig (which is a superset of .editorconfig) files and their `global_level` property. This is what we also do in our [Lombiq .NET Analyzers project](https://github.com/Lombiq/.NET-Analyzers) to have cascading configuration. E.g., [this file with `global_level = 0`](https://github.com/Lombiq/.NET-Analyzers/blob/dev/Lombiq.Analyzers/Lombiq.Analyzers.globalconfig) has the base configuration, and [this other one with `global_level = 10`](https://github.com/Lombiq/.NET-Analyzers/blob/dev/Lombiq.Analyzers.OrchardCore/Lombiq.Analyzers.OrchardCore.globalconfig) overrides some of it.

That works for defining code analysis rule settings, but does not apply to the text editor settings (tabs vs spaces, indent size, new lines/braces, etc.) which is what we're trying to solve here.

>Microsoft don't even follow the official specs in VS, especially trim_trailing_spaces.

Are you sure? It seems like whenever I press enter all the extra whitespace on my current line is trimmed before the editor moves to the next line. Is it possible you're expecting this behavior to apply to all lines in a file on save? I'm not aware of the spec defining the behavior of saving a file, only the behavior of actively editing a file (though admittedly I haven't spent much time with the spec). My curly brace behaviors apply when I create a curly brace, but it doesn't reformat all of the braces in a file if I manually move them to the previous line. Similarly, when I press tab it inserts spaces, but when a file already has tabs it doesn't replace them with spaces unless I manually tell it to. I believe one of my coworkers uses a setting that automatically formats the entire document on save (though this can be somewhat annoying at times).

>And the VSCode EditorConfig plugin is not maintained by Microsoft

Interesting, I didn't realize that. Is the VSCode functionality extensible? For example, can the C# Dev Kit for VSCode add .NET-specific functionality to it?

**Comment by cxw42** [comment]
> working across multiple projects with the same tooling / stack.

EditorConfig by design assumes nothing about relationships between projects.  (@treyhunner @xuhdev correct me if I'm wrong!)  I think that may be part of the communication challenge here.  

I am hearing two recurring themes in the comments:
* "It is easy to add this feature to the spec --- please do so."
* "It is easy to copy a file when you set up a project --- please do so."

:smile:  Both of those points are supportable, as witness the good arguments on both sides in this issue.

My own preference is to resolve #415 and then revisit #482 in view of whatever decisions are made in #415.  I think part of the strength of EditorConfig is its simplicity.  I think it is worth keeping things simple for use cases that don't need the complexity of features such as config inheritance.

**Comment by xuhdev** [comment]
I agree there's no assumption of relationship between projects here. Allowing referring to an external file would complicate the core libraries quite a bit. (It may not sound so complicated, but a change of requirement to core library would require changes in at least 6 core libraries with totally different implementations.) If we allow this, we would at least have a way to make sure the changes are compatible with the core libraries that do not implement this feature.

**Comment by marcospgp** [comment]
Also ran into this while trying to set up an `.editorconfig` that I can reuse across projects.

Ideally I could include an `.editorconfig` from a subfolder into my project's root `.editorconfig`, that way I could have a repo as the source of truth for `.editorconfig` and include it as a submodule in other repos.

Maybe I can also make a symbolic link for `.editorconfig` from the submodule folder to the root folder 🤔

**Comment by Janek91** [comment]
Hi, since this feature would be appreciated, what's the current status, or known workaround to achieve that?

**Comment by cxw42** [comment]
@Janek91
> workaround

Copy .editorconfig into each project

**Comment by cxw42** [comment]
I thought about this some more.  I think the closest analogy to my understanding of this request is CSS:

* CSS User-agent styles ~= individual editors' defaults
* CSS author styles ~= .editorconfig files from outside a repo (under discussion here)
* CSS user styles ~= .editorconfig files inside a repo
* (and we might have a fourth layer if people set up their editors to override the plugin)

I am not suggesting adopting CSS.  However, I do think it's a good source of insight into how others have solved what I perceive to be a similar problem.

**Comment by jackholden** [comment]
+1 for this feature. Love what @cxw42 has proposed. 

It would be extremely welcome, ensuring single source of truth. Very useful when not using a monorepo.

**Comment by cxw42** [comment]
An example of something analogous to CSS user styles: https://github.com/editorconfig/editorconfig/issues/220#issuecomment-3118132322

**Comment by andriy-ah** [comment]
@treyhunner @xuhdev @ffes 
Are there plans to ever implement this? This will be super useful for the monorepo.
Are there any workarounds we can use for the monorepo now excepy copy&pasting files and keeping them in sync?

**Comment by xuhdev** [comment]
> [@treyhunner](https://github.com/treyhunner) [@xuhdev](https://github.com/xuhdev) [@ffes](https://github.com/ffes) Are there plans to ever implement this? This will be super useful for the monorepo. Are there any workarounds we can use for the monorepo now excepy copy&pasting files and keeping them in sync?

For monorepos, you can put an `.editorconfig` file at the project root directory, and add an `.editorconfig` file in each subproject root to override settings.

**Comment by andriy-ah** [comment]
@xuhdev I am dealing with a massive monorepo and I don't open the entire repo at once, I open some specific projects/packages in it separately.

**Comment by xuhdev** [comment]
@andriy-ah What do you mean when you said opening the project? The `.editorconfig` file only need to sit in the filesystem. You don't have to open the entire repo in your IDE.

**Comment by tom-ricci** [comment]
Would love to be able to do this since I use Prettier in a turborepo monorepo and (by design) it only searches for EditorConfig files in the project root, so the only way to have Prettier apply EditorConfig options is to copy the EditorConfig file into each individual workspace that contains a Prettier configuration

**Comment by woldaker** [comment]
Also ran into this recently. Would love to see this. As of now, it's impossible to have two .editorconfig files that each have the same precedence and the same scope because there can only be one name: .editorconfig. And without being able to extend/include other .editorconfig files, the only way to do this is by putting one into a parent directory and one into a child directory. But what if we don't want to do that?

I have a situation where I would like to commit to git a core set of rules that will apply to an entire repo, but then I also want a set of personal rules that I will NOT commit to the repo but will also apply at the same level and precedence, i.e. I would need them to exist in the same directory.

**Comment by cxw42** [comment]
@woldaker I certainly don't blame you for wanting that.  However, having your own personal rules not shared with your team is, well, the opposite of what EditorConfig is designed to do 😅 .  The very first sentence on editorconfig.org says this project is for "consistent coding styles for multiple developers working on the same project".

**Comment by mark-wilkinson-dct** [comment]
+1 for this, we have a microservices architecture and if changes are required to our source editorconfig, it's a copy + paste job across dozens of repos. Being able to extend a properly versioned one (possibly via a locally distributed nuget package or something similar) would be a great solution

**Comment by kurnakovv** [comment]
Our team ran into a problem: `.editorconfig` has many limitations that prevent it from working properly in large .NET projects. For example, it's hard to conveniently and temporarily disable rules, it's impossible to partially inject rules for inactive files (to avoid git conflicts), it's impossible to split a file into multiple files (this is especially true when you have more than 1,000 lines of rules), and it's impossible to import existing rules without duplicating them in every project (which is exactly what this issue is about).

I created an open-source .NET tool called [biak](https://github.com/kurnakovv/biak) that solves all these problems. The idea is to create a wrapper around the `.editorconfig` file and work within it.

Below we will look at how to solve the problem related to this issue.

## Quick Start
* Install .NET version 6 or later
* Install biak
    ```
    dotnet tool install --global kurnakovv.biak
    ```
* Set up the project once
    ```
    dotnet biak setup
    ```

* Now you need to change the contents of the `.editorconfig-main` file with the necessary imports, for example

    .editorconfig-main
    ```.editorconfig
    root = true

    # Custom rules
    [*]
    insert_final_newline = true
    indent_style = space
    indent_size = 4
    trim_trailing_whitespace = true

    # Code analysis (CAxxxx) rules
    ^biak^ import https://gist.githubusercontent.com/kurnakovv/7ca966f8099209136f4baa87d804d39c/raw/cfb32e5b1da5c9dcd7340de5f1d166109642979d/.editorconfig

    # Code-style (IDExxxx) rules
    ^biak^ import https://gist.githubusercontent.com/kurnakovv/b396696805339988a547c6d53a0d0254/raw/193bb9d95f8132a690d7a3e51aa8291bd2bb4dba/.editorconfig

    # StyleCop.Analyzers rules
    ^biak^ import https://gist.githubusercontent.com/kurnakovv/cd01734d1775a1d53951d106f9ecf5b7/raw/a171d5f8bd686975e31354a2e419253096c803ef/.editorconfig

    # Roslynator rules
    ^biak^ import https://gist.githubusercontent.com/kurnakovv/ff955cd224602b738f697f3246fcda92/raw/1646b6563a79c4d2e78ae654152a3082c10428f6/.editorconfig
    ```

* The final step is to move the contents (with the imported rules) from `.editorconfig-main` to `.editorconfig`
    ```
    dotnet biak enable
    ```

    The contents of `.editorconfig` should now look something like this
    ```.editorconfig
    # ⚠️ Attention ⚠️
    #
    # This file is managed by the "biak" tool.
    # All changes are generated from the ".biak" folder.
    # Do not edit this file manually - your changes will be overwritten.
    # To modify its contents, update the sources inside ".biak" folder.
    #
    # To learn more about biak, visit: https://github.com/kurnakovv/biak

    root = true

    #Custom rules
    [*]
    insert_final_newline = true
    indent_style = space
    indent_size = 4
    trim_trailing_whitespace = true

    # Code analysis (CAxxxx) rules
    ...

    # Code-style (IDExxxx) rules
    ...

    # StyleCop.Analyzers rules
    ...

    # Roslynator rules
    ...

    # ⚠️ Attention ⚠️
    #
    # This file is managed by biak.
    # Do not edit it manually - your changes will be overwritten.
    # See the note at the top ☝️ of the file for details.
    ```

    Where `...` will be replaced with the content from the links.

    Now, when changing content in `.editorconfig-main` or in the links, you need to run the `dotnet biak enable` command.

Also, if you simply need to split a large file into multiple files, you can use imports for the same purpose.

Structure
```
.biak/
├── Categories/
│ ├── .editorconfig-CodeAnalysis-CAxxxx
│ ├── .editorconfig-CodeStyle-IDExxxx
│ ├── .editorconfig-IDisposableAnalyzers
│ ├── .editorconfig-kuker
│ ├── .editorconfig-Roslynator
│ └── .editorconfig-StyleCop
│
├── .editorconfig-main
└── config.json
```
> [!WARNING]
> It's important to place everything inside the `.biak` folder, as nothing outside of it will work. This is done for security purposes, to prevent content from being pulled in from outside, for example, via GitHub actions.

.editorconfig-main
```.editorconfig
root = true

[*]
insert_final_newline = true
indent_style = space
indent_size = 4
trim_trailing_whitespace = true

^biak^ import "Categories/.editorconfig-CodeAnalysis-CAxxxx"

^biak^ import "Categories/.editorconfig-CodeStyle-IDExxxx"

^biak^ import "Categories/.editorconfig-IDisposableAnalyzers"

^biak^ import "Categories/.editorconfig-kuker"

^biak^ import "Categories/.editorconfig-Roslynator"

^biak^ import "Categories/.editorconfig-StyleCop"
```

And again, you need to run the `dotnet biak enable` command.

You can read more about imports in the [docs](https://github.com/kurnakovv/biak/wiki/Import)

I understand that not everyone here uses .NET, and perhaps self-contained binaries are worth considering. I created issue [#82](https://github.com/kurnakovv/biak/issues/82). If this is useful, please provide feedback.

Yes, it's an additional layer on top of `.editorconfig`, and it would be great to work without it, but unfortunately this issue has been open since 2015 and still hasn't been addressed, even though many people need it.

I hope this tool is useful to someone ;)

**Comment by florianb** [comment]
> Would love to be able to do this since I use Prettier in a turborepo monorepo and (by design) it only searches for EditorConfig files in the project root, so the only way to have Prettier apply EditorConfig options is to copy the EditorConfig file into each individual workspace that contains a Prettier configuration

Thanks @tom-ricci - Prettier does not support the out of project traversal ("Unlike the EditorConfig spec, the search for .editorconfig file will stop on the project root and won't proceed further.") most probably for the following reason:

> Prettier intentionally doesn’t support any kind of global configuration. This is to make sure that when a project is copied to another computer, Prettier’s behavior stays the same. Otherwise, Prettier wouldn’t be able to guarantee that everybody in a team gets the same consistent results. (see [here](https://prettier.io/docs/configuration))

We already support this behaviour for properties which are not explicitly defined by a subfolder.

I really doubt Prettier would add any support for another referencing property.

---

As for big multirepo projects - i get it's hard to manually copy editorconfigs but this is still an explicit change to the project, please use a script to lower the maintenance burden.

@mark-wilkinson-dct - thanks a lot - did you consider adding an upper directory as a project holding a central `.editorconfig`?

**Comment by cxw42** [comment]
@kurnakovv thanks for sharing your tool!  Automating the generation of EC files for your specific environment seems perfectly reasonable to me.

**Comment by TheXenocide** [comment]
I could really use any form of import at this point. If I could so much as put a before.editorconfig and after.editorconfig in my repo-root to live next to an identical base template that gets automatically updated in many downstream repos, this would be enough for my use case to be able to apply repo-specific/project-specific overrides in a predictable manner while sharing the vast majority of my configuration. I was really hoping to avoid the copy/paste-ness of checking in duplicate files by leveraging my platform's native packaging mechanisms, but I can see how that would be quite the burden to design in a way that satisfies all the other platforms. 

If the easiest path works well enough I could make do with it. This conversation has been sitting here for years and I could use basically any answer to merging a nearly identical file to 50 repos or forcing users to run scripts on clean checkouts just to setup an environment that is supposed to make collaboration easier/more consistent; this kind of thing "just working" out of the box is a boon to us who want to receive the contributions more consistently.

**Comment by kurnakovv** [comment]
@TheXenocide If you want, you can try [biak](https://github.com/kurnakovv/biak). It supports link imports, and you don't have to duplicate the same rules, but rather store them in one place (for example, by uploading them to a GitHub gist).
I wrote about this option [above](https://github.com/editorconfig/editorconfig/issues/236#issuecomment-4363349107).

If you've already seen this method and it doesn't work for you, I apologize.
